compiler/src/iree/compiler/Codegen/Common/GPU/GPUApplyTilingLevel.cpp

```diff
@@ -0,0 +1,203 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/GPU/Passes.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenInterfaces.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.h"
+#include "llvm/ADT/DenseSet.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/TileUsingInterface.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Interfaces/TilingInterface.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-tensor-tile"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_GPUAPPLYTILINGLEVELPASS
+#include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
+
+namespace {
+struct GPUApplyTilingLevelPass final
+    : impl::GPUApplyTilingLevelPassBase<GPUApplyTilingLevelPass> {
+  using GPUApplyTilingLevelPassBase::GPUApplyTilingLevelPassBase;
+  void runOnOperation() override;
+};
+} // namespace
+
```

Comment:
Add some comment for this function like collecting (including) the given op's producers that are tilable and fusable